// gatsby-plugin-page-load-delay/gatsby-browser.js
'use strict';

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createBrowserHistory = require('history/createBrowserHistory');

var _createBrowserHistory2 = _interopRequireDefault(_createBrowserHistory);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Change history
var historyExitingEventType = 'history::exiting';
var getUserConfirmation = function getUserConfirmation(pathname, callback) {
	if (pathname === document.location.pathname) return;
	var event = new CustomEvent(historyExitingEventType, { detail: { pathname: pathname } });
	window.dispatchEvent(event);
	// Delay if exit time has been previously supplied by a component
	setTimeout(function () {
		callback(true);
	}, window.pageExitTime);
	window.pageExitTime = 0;
};
var history = (0, _createBrowserHistory2.default)({ getUserConfirmation: getUserConfirmation });
history.block(function (location, action) {
	return location.pathname;
});
exports.replaceHistory = function () {
	return history;
};

var ReplaceComponentRenderer = function (_React$Component) {
	(0, _inherits3.default)(ReplaceComponentRenderer, _React$Component);

	function ReplaceComponentRenderer(props) {
		(0, _classCallCheck3.default)(this, ReplaceComponentRenderer);

		var _this = (0, _possibleConstructorReturn3.default)(this, (ReplaceComponentRenderer.__proto__ || (0, _getPrototypeOf2.default)(ReplaceComponentRenderer)).call(this, props));

		_this.state = { status: 'entered' };
		_this.listenerHandler = _this.listenerHandler.bind(_this);
		return _this;
	}

	(0, _createClass3.default)(ReplaceComponentRenderer, [{
		key: 'listenerHandler',
		value: function listenerHandler(event) {
			this.setState({ status: 'exiting' });
		}
	}, {
		key: 'componentDidMount',
		value: function componentDidMount() {
			window.addEventListener(historyExitingEventType, this.listenerHandler);
		}
	}, {
		key: 'componentWillUnmount',
		value: function componentWillUnmount() {
			window.removeEventListener(historyExitingEventType, this.listenerHandler);
		}
	}, {
		key: 'componentWillReceiveProps',
		value: function componentWillReceiveProps(nextProps) {
			var _this2 = this;

			if (this.props.location.key !== nextProps.location.key) {
				this.setState({ status: 'entering' });
				setTimeout(function () {
					_this2.setState({ status: 'entered' });
				}, 1);
			}
		}
	}, {
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'div',
				null,
				(0, _react.createElement)(this.props.pageResources.component, (0, _extends3.default)({}, this.props, this.props.pageResources.json, {
					status: this.state.status
				}))
			);
		}
	}]);
	return ReplaceComponentRenderer;
}(_react2.default.Component);

exports.replaceComponentRenderer = function (_ref) {
	var props = _ref.props,
	    loader = _ref.loader;

	if (props.layout) return undefined;
	return (0, _react.createElement)(ReplaceComponentRenderer, (0, _extends3.default)({}, props, { loader: loader }));
};
